### ISSUE TITLE
Better handling of recursion in `objectBuilder` directives

### ISSUE THREAD

**Issue by abensonca** [galacticusorg/galacticus#695]
It should be possible to detect and handle recursion during construction of objects from the parameters structure in an entirely automated way. 

In the `functionClass`-level parameters constructor:

1. Detect if we ever return to the same node in the parameter file
   - would need a stack (that keeps a pointer to the `inputParameter`) for this as could go through this class multiple times in an `objectBuilder`
 2. If detected, return a shim version of the class that just retains a pointer to the original instance and calls the function in the original. 

Would need to ensure that the pointer to the object being constructed that is set in the `inputParameter` node is set _prior_ to anything else in the construction process, so that we can reference this pointer as needed.

**Comment by abensonca** [comment]
# Issue #695 — Automatic handling of recursion in `objectBuilder` directives: review and plan

*Review date: 2026-07-05. All file/line references are to the current `master` (82ac5cb3d).*

## Summary

Issue #695 proposes that recursive construction of `functionClass` objects (an object whose parameter-driven construction cycles back to itself) be detected and handled automatically, instead of the current per-class manual mechanism. The review finds that **most of the proposed detection infrastructure already exists** — largely built for issue #397 (the recursion *guard*, merged 2026-06) and for the original 2019 `recursive="yes"` machinery. What remains genuinely new is:

1. **Auto-generating the shim** (the lightweight object that forwards all method calls to the instance already under construction) — today ~100–115 lines of subtle hand-written boilerplate in each of the three classes that support recursion.
2. **Auto-generating the deep-copy fix-up** for the shim's back-pointer — today wholly manual (`parentDeferred`/find-parent two-pass scheme) and historically race-prone.
3. **Unifying** the two currently-parallel subsystems (the #397 abort-on-cycle build stack, and the `recursive="yes"` short-circuit) into one mechanism.

One important design caveat emerged: **full automation (no opt-in at all) is not safe**, because some construction cycles are semantically divergent — auto-shimming them would convert today's clean construction-time error into an infinite *runtime* recursion. The recommendation is to keep `recursive="yes"` as a one-word opt-in that asserts "cycling back to me is semantically bounded", while eliminating *all* the per-class boilerplate behind it.

---

## 1. Current state

### 1.1 Two parallel subsystems

**Subsystem A — the recursion guard (issue #397).** Detects an *accidental* unbounded recursive build and aborts with a diagnostic. Components:

- A thread-private build stack in `source/utility/input_parameters.F90:275-382`. Each `buildStackEntry` holds a pointer to the `inputParameter` node being built, the class name, and a `recursionAware` flag. `Input_Parameters_Build_Stack_Push` (`:309-369`) scans for a repeat of the same node+class; if none of the intervening stack entries is recursion-aware, it aborts with the full build-stack trace (`:339-352`).
- Every generated class-level factory brackets its constructor dispatch with push/pop, threading `recursionAware = .true.` iff any class in the hierarchy is `recursive="yes"` (`python/Galacticus/Build/SourceTree/Process/FunctionClass/__init__.py:3179-3217`).
- A per-`objectBuilder` same-class static guard emitted when `self` is in scope (`python/Galacticus/Build/SourceTree/Process/ObjectBuilder.py:189-201`).
- Pinned by `python/Galacticus/Build/SourceTree/tests/test_object_builder_recursion_guard.py` and exercised end-to-end by `testSuite/test-objectBuilder-recursion.py` (self-composition via `galacticFilterNot`; mutual recursion via `criticalOverdensityEnvironmental` ↔ `haloEnvironmentNormal`).

**Subsystem B — the manual `recursive="yes"` mechanism (2019).** Lets a class *legitimately* participate in a cycle. The generated half (`FunctionClass/__init__.py`):

- Thread-private module variables `<name>RecursiveBuildNode` / `<name>RecursiveBuildObject`, emitted iff any class in the family is recursive (`:2993-3004`).
- In the factory dispatch, for a recursive class these are pointed at the parameter node and at `self` **before** the inner constructor runs, and cleared after (`:3183-3202`). This already satisfies the issue's point 3 ("pointer to the object under construction set prior to anything else") — via module variables rather than the node's object cache. (The node cache, `inputParameter%objects` + `objectCreated`/`objectGet`/`objectSet` at `input_parameters.F90:69-106, 1382-1481`, is populated only *after* construction completes, in the objectBuilder template at `ObjectBuilder.py:226-231`.)
- The short-circuit: on re-entry for the *same node currently under construction*, the factory allocates a fresh instance of the concrete recursive type, sets `self%isRecursive=.true.` and `self%recursiveSelf => <name>RecursiveBuildObject`, and returns immediately (`:3144-3171`). That instance is the shim.
- Generated `stateStore`/`stateRestore` short-circuits that forward to `recursiveSelf` (`:2470-2492`).

The hand-written half, per class (the boilerplate this issue wants to eliminate):

- `recursive="yes"` on the directive, plus `<deepCopy><ignore variables="recursiveSelf"/></deepCopy>` and `<stateStorable><exclude variables="recursiveSelf"/></stateStorable>`.
- Type components `logical :: isRecursive, parentDeferred` and `class(<self type>), pointer :: recursiveSelf`, with constructor initialisation.
- An `if (self%isRecursive) then; call self%recursiveSelf%<method>(...); return; end if` prologue at the top of **every** public method (the shim holds no real state; missing one guard means operating on uninitialised components).
- A custom `deepCopy`/`deepCopyReset`/`deepCopyFinalize` trio plus `…DeepCopyAssign` and `…FindParent` helpers implementing a two-pass copy fix-up: the real object records where its copy went (repurposing its own `recursiveSelf` slot); a shim being copied either points its copy at the parent's recorded copy, or defers (`parentDeferred=.true.`) and resolves in `deepCopyFinalize` via find-parent. `deepCopyFinalize` was added in 2021 (commit `2ada7cb19`) specifically because the earlier scheme had cross-thread races.

### 1.2 Inventory of `recursive="yes"` classes

Exactly three classes opt in:

| Class | File | Cyclic member | Guarded methods | ~Boilerplate LOC |
|---|---|---|---|---|
| `starFormationHistoryFixedAges` | `source/star_formation/histories/fixed_ages.F90:28` | `geometryLightcone_` (lightcone geometry composites a `darkMatterHaloScale`, which can cycle back) | 10 | ~115 |
| `darkMatterHaloScaleVirialDensityContrastDefinition` | `source/dark_matter_halos/scales/virial_density_contrast.F90:31` | `virialDensityContrast_` | 9 | ~110 |
| `virialDensityContrastPercolation` | `source/structure_formation/virial_density_contrast/percolation/_class.F90:30` | `percolationObjects_` container → dark-matter-profile objects → `darkMatterHaloScale` | 2 (+ helpers) | ~100 |

The second and third form the canonical mutual cycle: `virialDensityContrastPercolation` → percolation objects → dark-matter profiles → `darkMatterHaloScaleVirialDensityContrastDefinition` → `virialDensityContrastPercolation`. Total: **~325 lines of hand-written recursion boilerplate**, plus notable warts:

- `percolationCopyTable` (`percolation/_class.F90:480-505`, table-sharing between a recursive child and its parent) is **defined but never called** — dead code.
- `virial_density_contrast.F90` omits the `<stateStorable><exclude variables="recursiveSelf"/>` block the other two carry — an inconsistency that automation would eliminate by construction.
- `virialDensityContrastPercolation` additionally carries a module-scope thread-private `solving` flag (`percolation/_class.F90:94-96, 240-273`) preventing re-entry during its tabulation loop. This is *domain* re-entry logic, not construction recursion — it stays regardless of this issue.

### 1.3 History

Mechanism introduced June 2019 (`ab0539dc2`, motivated by circular-reference memory leaks); halo-scale class June 2019; general factory-level detection September 2019 (`0c3c18e93`); `deepCopyFinalize`/find-parent race fix February 2021 (`2ada7cb19`); `starFormationHistoryFixedAges` added October 2024 (`524241c47`); the #397 guard with its `recursionAware` interplay June 2026 (`1b6d30b10`).

---

## 2. Assessment of the issue's proposal, point by point

> *1. Detect if we ever return to the same node in the parameter file — would need a stack (that keeps a pointer to the `inputParameter`) …*

**Already exists** — the #397 build stack is exactly this, including the "could go through this class multiple times legitimately" nuance (the `recursionAwareBetween` scan, `input_parameters.F90:330-338`). What it does *not* yet hold is a pointer to the object under construction (that lives in the separate per-family `<name>RecursiveBuildObject` variables), and its response to detection is abort rather than shim-return.

> *2. If detected, return a shim version of the class that just retains a pointer to the original instance and calls the function in the original.*

**Exists but is manual.** The factory short-circuit already returns exactly such a shim — but the shim is the concrete class itself with an `isRecursive` flag, and the "calls the function in the original" part is ~10 hand-written prologues per class. The genuinely new work is generating the shim automatically. The preprocessor has everything needed: `_generate_method_functions` (`FunctionClass/__init__.py:3718-3871`) already reconstructs the *full signature* of every method (arguments, return type, `recursive`/`elemental` prefixes) to emit default implementations, so emitting forwarding bodies instead is mechanical.

> *3. Would need to ensure that the pointer to the object being constructed that is set in the `inputParameter` node is set prior to anything else in the construction process.*

**Solved, differently.** The generated factory sets `<name>RecursiveBuildObject => self` before dispatching the inner constructor (`__init__.py:3187-3191`). Storing the in-progress pointer *in the node itself* (via `objectSet`) would be a behaviour change — the node cache currently means "construction complete, reusable", and the objectBuilder reuse path (`ObjectBuilder.py:206-216`) increments the reference count on it, which must not happen for an in-progress object. Recommendation: keep "in-progress" state out of the node cache and instead carry the object pointer on the build stack (see D3 below), which also removes the per-family module variables.

> *(comment) The shim class could probably then be handled automatically during `deepCopy` also.*

**Correct, and this is the largest piece of new work.** Today there is zero generated awareness of `recursiveSelf` in deepCopy — the generator only honours the `<ignore>` directive; the entire two-pass fix-up is hand-written per class (§1.1). Automating it is what removes the most fragile boilerplate.

---

## 3. Key design decisions

### D1 — Fully automatic vs. retained opt-in. **Recommend: keep `recursive="yes"` as the opt-in, eliminate everything else.**

Not every construction cycle is semantically valid. The #397 test cases show why: a `galacticFilterNot` whose child resolves back to itself would, under automatic shimming, construct "successfully" — and then diverge at *call* time (`passes()` → child shim → forwards to the same filter → `passes()` → …). Automatic shimming everywhere would convert today's clean construction-time diagnostic into a runtime hang/stack overflow with no message — a strict regression for user error handling, and #397 was implemented precisely because those crashes were painful to diagnose. There is no static way to distinguish "bounded" cycles (the legitimate three, where the physics guarantees the recursion terminates) from divergent ones.

The opt-in should shrink to *just the directive attribute*: one word asserting "re-entry into me is semantically bounded". All the machinery behind it — shim, method forwarding, deepCopy, state-store, descriptor handling — becomes generated. Non-opted-in classes keep the #397 abort. This can be revisited later (e.g. an opt-out) once the automated mechanism has burned in; the #397 diagnostic could also be extended to mention `recursive="yes"` as the resolution when the cycle passes through a class the developer controls.

### D2 — Shim representation. **Recommend: a single generated shim type per functionClass family.**

Two candidate designs:

**(a) Generated family-level shim type** — emit, per functionClass directive (or only for families containing a recursive class), a `type, extends(<name>Class) :: <name>Recursive` holding only the weak `recursiveSelf => class(<name>Class)` pointer, with every deferred method implemented as a generated forwarder. The factory short-circuit allocates this type instead of the concrete class.

**(b) Generalise the status quo** — keep allocating the concrete type with an `isRecursive` flag, and have the preprocessor *inject* the forwarding prologue into every hand-written type-bound method body.

Option (a) is strongly preferable:

- No per-implementation code generation or injection into hand-written method bodies (injection is fragile: private helpers, methods bound under different names, code the parser mis-scans).
- The shim physically *cannot* touch uninitialised concrete-class components — the "missed one guard" bug class disappears.
- deepCopy/descriptor/stateStore/allowedParameters/autoHook for the shim are written once, centrally, in the generator.
- It automatically covers **every implementation** in the family, not just the specific class that today carries the boilerplate.

Risks of (a), and mitigations:

- **`select type` on concrete types.** Code doing `type is (<concreteImpl>)` on an object obtained via objectBuilder would not match a shim. Survey result: *no* `type is` on any of the three current recursive concrete types outside their own files, so the migration itself is safe; a family-wide audit should be part of the work for any future opt-ins. (Note the status quo has the mirror-image problem anyway: a shim of the concrete type *matches* `type is` but has uninitialised components — arguably worse.)
- **Elemental methods.** A forwarding body inside an `elemental` procedure references a pointer component; needs a compile-time check across all method attribute combinations (the generator already knows each method's prefixes at `__init__.py:3772-3777`).
- **Generic bindings / optional arguments.** Forwarders must pass optional arguments through verbatim (safe in Fortran when forwarding to a matching signature) and re-declare `generic ::` interfaces; the generator already has this metadata (`:2902-2908`).

### D3 — Unify detection into the build stack. **Recommend: extend `buildStackEntry` with the object-under-construction pointer.**

Add `class(functionClass), pointer :: object` to `buildStackEntry`, set by the factory before dispatching the inner constructor (where `<name>RecursiveBuildNode/Object` are set today). Then:

- On push, when a repeat of (node, className) is found: if the matching entry's family allows recursion → return the shim wired to that entry's object; else → the existing #397 abort. One mechanism, one code path, one stack.
- The per-family `<name>RecursiveBuildNode`/`<name>RecursiveBuildObject` thread-private module variables (`__init__.py:2996-3004`) are deleted — the stack subsumes them.
- The stack is already thread-private, so the OpenMP story is unchanged.

One subtlety to preserve: today's short-circuit fires when re-entering the node *currently* under construction; the stack must return the object from the *matching entry* (the same node+class), not merely the innermost recursive entry, so deep cycles that pass through the class twice wire the shim to the right instance — this is exactly the information the stack already indexes.

### D4 — Automate deepCopy of the shim. **Recommend: generate the two-pass fix-up centrally; evaluate reusing `copiedSelf`.**

With a dedicated shim type, its deepCopy is generated once per family:

1. Copying a shim: allocate a new shim; if the pointed-to real object was already copied in this operation, point at its copy; otherwise keep the original pointer and mark deferred.
2. `deepCopyFinalize` pass: resolve deferred shims to the copied parent.

The existing hand-written scheme repurposes the *real* object's `recursiveSelf` slot to record where its copy went. But every generated `<name>Class` already carries a public `copiedSelf` pointer for exactly this purpose (`__init__.py:2820-2823`), maintained by the generated deepCopy/reset/finalize (`:2212-2215`). The shim's deferred resolution can likely read `recursiveSelf%copiedSelf` in `deepCopyFinalize` instead of a bespoke recording mechanism — eliminating `parentDeferred` gymnastics and the repurposed-slot trick entirely. **Caveat:** the 2021 fix (`2ada7cb19`) exists because cross-thread copies once resolved to the wrong thread's parent; `copiedSelf` reset/finalize ordering within one deepCopy operation must be audited before committing to this (specifically: `copiedSelf` is nulled in `deepCopyFinalize`, so shim resolution must run before parents are finalized, or the finalize ordering must be two-staged). If the audit says no, generate the current record-and-fix-up scheme instead — still centrally, still zero per-class code.

`deepCopyReset` for a shim: reset its own copy-bookkeeping but never null its `recursiveSelf` (that is its only real state).

### D5 — Per-facility shim semantics

The generator must define, once, what a shim does for each generated facility:

| Facility | Shim behaviour | Notes |
|---|---|---|
| All class methods | Forward to `recursiveSelf` | Generated from full signature metadata (`__init__.py:3718-3871`). |
| `stateStore`/`stateRestore` | Forward | Pattern already generated for `recursive="yes"` classes (`:2470-2492`); moves into the shim type. |
| `descriptor`/`hashedDescriptor` | Forward | Currently hand-guarded in `fixed_ages` only; **no** generated `isRecursive` awareness exists — an untreated gap today. Forwarding terminates because the real object's descriptor doesn't re-enter the shim (its parameters serialise the original sub-tree). Verify no cycle via the child that owns the shim. |
| `allowedParameters` | No-op (or forward) | The shim holds no child objects; today this works by accident (unassociated pointers in a concrete-type shim). Make it explicit. |
| `autoHook` | **No-op** | The real object already hooked itself; the objectBuilder template calls `autoHook` on every built object (`ObjectBuilder.py:232`), so a forwarding or default autoHook would double-register on `calculationResetEvent`. Today's manual shims *do* attach (e.g. `virial_density_contrast.F90:153-163`) and then forward `calculationReset` — the generated design should instead not hook at all; decide and document. |
| `calculationReset` | Forward (if not hooked) or no-op (if the real object receives the event itself) | Must be consistent with the `autoHook` decision. |
| Reference counting | Shim itself refcounted normally by its owner; `recursiveSelf` is a **weak** reference — never incremented, never decremented, never destroyed through | Anything else creates an ownership cycle: ancestor → child → shim → ancestor never reaches zero and leaks. This is precisely the 2019 motivation (`ab0539dc2`) and must be a hard rule in the generated destructor. |
| `objectDestructor` on the shim | Deallocate the shim only; leave `recursiveSelf` untouched | Lifetime is safe by structure: the real object transitively owns the shim, so it outlives it — but deepCopy/teardown ordering should be asserted in debug builds (e.g. error if a shim outlives its parent, mirroring today's `"recursive child's parent was not copied"` check). |

---

## 4. Test-coverage gaps (pre-existing)

- **No targeted test exercises a *successful* recursive build.** Both #397 tests pin the *abort* path. The legitimate path is only covered incidentally: `virialDensityContrastPercolation` via the three `test-mass-conservation-*` parameter files, `starFormationHistoryFixedAges` via the lightcone regression tests (`testSuite/regressions/lightconeCrossing*.xml`, `test-star-formation-histories-fixedAge.xml`).
- **No test of deepCopy of a recursive object graph** (the most fragile part, with a real race-condition history), nor of stateStore/descriptor on shims, nor an OpenMP-threaded recursive build.

These gaps must be closed *before* refactoring (Phase 0 below), since the migration rewrites exactly the code they would protect.

---

## 5. Implementation plan

**Phase 0 — Characterisation tests (no production changes).**
Add a testSuite case with a parameter file that triggers the legitimate recursive build (percolation ↔ halo-scale cycle is the natural one; optionally also `fixedAges` + lightcone) and asserts a successful run; extend it (or add a unit-style test) to force a deepCopy of the recursive structure (e.g. a multi-thread/`evolveForests` configuration) and a stateStore/descriptor round-trip. Add generator-level pin tests alongside `test_object_builder_recursion_guard.py` for the short-circuit block. *Deliverable: green tests on master pinning today's behaviour.*

**Phase 1 — Unify detection (behaviour-preserving).**
Extend `buildStackEntry` with the object-under-construction pointer; factories push it before dispatching the inner constructor for all classes. Reimplement the `recursive="yes"` short-circuit as a stack lookup (return shim wired to the matching entry's object) and delete the `<name>RecursiveBuildNode`/`<name>RecursiveBuildObject` module variables and the `recursionAware` in-between scan (the stack entry itself now knows whether its family allows recursion). Non-recursive families keep the #397 abort verbatim. Update the #397 pin tests for the new identifiers. *Deliverable: identical runtime behaviour, one detection mechanism.*

**Phase 2 — Generate the shim type.**
For each family containing a `recursive="yes"` class, emit `type, extends(<name>Class) :: <name>Recursive` with: weak `recursiveSelf` pointer; generated forwarders for every method (reusing `_generate_method_functions` signature reconstruction); generated `deepCopy`/`deepCopyReset`/`deepCopyFinalize` per D4; `stateStore`/`stateRestore`/`descriptor` forwarding; `allowedParameters` and `autoHook` no-ops per D5. Switch the factory short-circuit to allocate `<name>Recursive`. The old concrete-type-with-flag path remains temporarily for the three classes (both mechanisms coexist behind the attribute). *Deliverable: shim generation exercised by Phase-0 tests against at least one migrated family.*

**Phase 3 — Migrate the three classes and delete the boilerplate.**
Remove from all three: `isRecursive`/`parentDeferred`/`recursiveSelf` components, constructor initialisations, every per-method guard, the custom deepCopy trio + `DeepCopyAssign`/`FindParent` helpers, the `<deepCopy><ignore>`/`<stateStorable><exclude>` blocks, and the dead `percolationCopyTable`. Keep percolation's `solving` flag (domain logic) and decide whether its table-sharing intent (never actually wired up) should be recreated properly or formally dropped. Remove the now-unused `isRecursive` branches from the stateStore generator and the legacy `recursiveConstruct, recursiveSelf` constructor-signature tolerances (`__init__.py:720-723, 1614-1618`) — checking `testSuite/test-python-utils.py` alongside `pytest python/`, which is known not to cover it. *Deliverable: ~325 lines of manual boilerplate deleted; Phase-0 tests still green; full testSuite pass.*

**Phase 4 — Documentation and diagnostics.**
Document `recursive="yes"` (one attribute, everything else automatic) in the developer docs; extend the #397 abort message to suggest the attribute when the cycle passes through a single class family; note the runtime-divergence rationale for keeping the opt-in (D1) in the issue when closing.

**Suggested audits during Phase 2** (cheap, avoids surprises): grep-audit for `type is (` / `same_type_as` against implementations of the affected families; compile-check forwarders for `elemental` methods across all families before generalising beyond the three.

## 6. Effort and risk estimate

- Phase 0: small (one parameter file + test script + generator pin tests).
- Phase 1: moderate — touches `input_parameters.F90` and `_generate_constructor`; behaviour-preserving so well-protected by Phase 0.
- Phase 2: the bulk of the work — new generation code in `FunctionClass/__init__.py`; the deepCopy fix-up (D4) is the riskiest element given the 2021 race history; the `copiedSelf`-reuse audit decides between two implementations.
- Phase 3: mostly deletion; low risk once Phase 2 is proven.
- Main residual risks: deepCopy ordering across threads (D4 caveat), `autoHook`/`calculationReset` double-registration semantics (D5), and elemental-method forwarding (D2) — each has a specific audit called out above.

**Comment by abensonca** [comment]
# Addendum: destruction and memory-leak analysis (relation to #119)

A follow-up question to the review above: do the recursive cases need anything special to avoid memory leaks, and how is destruction handled now?

## Short answer

No extra machinery is needed — *provided the automated design preserves the one property that makes the current scheme work*: `recursiveSelf` is a **weak reference**. It is never `referenceCountIncrement`ed (not in the generated short-circuit at `FunctionClass/__init__.py:3144-3171`, not in the hand-written classes), never decremented, and never destroyed through. That is what breaks the ownership cycle so plain reference counting stays sound. This was the original motivation for the mechanism — commit `ab0539dc2` (2019) introduced it "to avoid circular references between functionClass objects which can lead to memory leaks". In other words, the `recursiveSelf` system *is* the weak-reference approach suggested in the comment on #119, applied at the one place parameters-driven construction can legally form a cycle.

## How destruction works today

All ownership references are counted: the `objectBuilder` template increments on build and on cache reuse (`ObjectBuilder.py:215, 230`), the parameter-node cache increments when it stores an object (`input_parameters.F90:1476`), and `objectDestructor` decrements and deallocates at zero (`ObjectBuilder.py:424-441`). The node cache itself is released in `inputParameterReset` (`input_parameters.F90:1500-1518`), which runs an `objectDestructor` on the cached object when the parameters are reset or destroyed.

For a recursive structure (percolation example: `A` → percolation objects → profile → halo scale `S` → shim(`A`)): destroying `A` runs its finalizer, which `objectDestructor`s its children → `S` reaches zero → `S`'s finalizer destroys the shim. The shim's finalizer is safe because its child pointers are default-initialized to `null()` (`fixed_ages.F90:50-52`), so the `if (associated(...))` guards in the shared finalizer all no-op, and nothing touches `recursiveSelf`. Since the shim's lifetime is nested inside `A`'s subtree, the weak pointer cannot dangle during teardown. No leak, no double-free; reference counting suffices.

## Two warts found, worth folding into the implementation plan

**1. The shim can permanently hijack the parameter-node object cache.** While the outer build of node `N_A` is in progress, the inner `objectBuilder` that re-discovers `N_A` receives a shim from the factory and — because the template unconditionally does `call parameterNode%objectSet(shim)` — stores the *shim* in `N_A`'s cache. When the outer build finishes, its `objectBuilder` calls `objectSet(A)`, but the slot is already occupied and `objectSet` silently skips (`input_parameters.F90:1473`). Consequences: the real object is cached nowhere; every future requester of that node receives the shim (functionally correct — it forwards — but a permanent extra indirection); and there is a latent use-after-free: the node holds a counted reference keeping the shim alive, while the shim's weak `recursiveSelf` does nothing to keep `A` alive, so if `A`'s owners release it a later `objectGet` returns a shim with a dangling parent. Benign in practice today (these objects are run-lifetime defaults), but the automated design should fix it: have the `objectBuilder` skip `objectSet` for shims (trivially identifiable once there is a generated `<name>Recursive` type). That restores "real object in the cache" and gives a structural guarantee that the weak pointer is always nested inside its target's lifetime. (Suggested home: Phase 2 of the plan above.)

**2. Double event-hooking.** The `objectBuilder` calls `autoHook()` on whatever the factory returns — including shims. The halo-scale shim therefore attaches itself to `calculationResetEvent` (`virial_density_contrast.F90:161`) alongside the real object, and its `calculationReset` forwards, so the real object is reset twice per event. Harmless (resets are idempotent) but wasteful, and it forces the shim's destructor to carry the `isAttached`/`detach` dance (`:179`). This confirms the D5 recommendation that the generated shim's `autoHook` should be a no-op.

## Relation to #119

The automation proposed here does not create any new need for garbage collection — the opposite: with #397 aborting unintended cycles, and the shim's weak back-edge breaking intended ones at the only point parameters-driven construction can form a loop, reference counting remains sound for everything built from a parameter file. What #119 would still cover is cycles wired up *outside* the parameters path (internal constructors connecting objects manually), which construction-time shimming cannot address. If anything, generating the shim (and its trivially-correct destructor) removes the "complicated system of `recursiveSelf` pointers" that #119 identifies, weakening the case for a full GC.

A cheap belt-and-braces option for the plan: give the real object a live-shim counter (incremented at shim creation, decremented in the generated shim destructor) and report an error in the real object's finalizer if it is nonzero — turning any future dangling-weak-reference bug into a clean diagnostic instead of a use-after-free.

**Comment by abensonca** [comment]
# Implementation summary

Implemented on branch `feature/issue-695-auto-recursion` (off `master`), following the plan above. The recursion machinery is now generated behind a single `recursive="yes"` opt-in; all of the per-class hand-written boilerplate is gone.

## What changed

**Phase 0 — characterization tests.** Added a targeted test of a *successful* recursive build (`test-objectBuilder-recursion-success.py` + `recursionSuccessPercolation.xml`, the percolation ↔ halo-scale cycle), run multi-threaded to force a deepCopy of the recursive object graph and, via `stateFileRoot`, a stateStore round-trip; plus generator-level pin tests. These closed the pre-existing coverage gaps (no test previously exercised a successful recursive build, a deepCopy of a recursive graph, or stateStore/descriptor of a shim) before any production change.

**Phase 1 — unify detection (behaviour-preserving).** `buildStackEntry` gained a weak `class(*), pointer :: object`; two helpers (`Input_Parameters_Build_Stack_Object_Set`/`_Recursive_Object`) record/query the object under construction on the shared #397 build stack. The per-family thread-private `<name>RecursiveBuildNode`/`RecursiveBuildObject` module variables are deleted — the short-circuit now queries the stack. The #397 abort path is untouched.

**Phase 2 — generate the shim type.** For each family with a `recursive="yes"` class, the preprocessor now emits `type, extends(<name>Class) :: <name>Recursive`, holding only the weak `recursiveSelf` back-pointer and forwarding every method (plus `stateStore`/`stateRestore`/`descriptor`/`objectType`) to it; `allowedParameters` and `autoHook` are no-ops; `deepCopy`/`deepCopyReset`/`deepCopyFinalize` are generated centrally.

**Phase 3 — migrate all three classes, delete the boilerplate.** The factory short-circuit now allocates the shim for every recursive family. Removed from `starFormationHistoryFixedAges`, `darkMatterHaloScaleVirialDensityContrastDefinition`, and `virialDensityContrastPercolation`: `isRecursive`/`parentDeferred`/`recursiveSelf` components, per-method forwarding guards, custom deepCopy trios + `DeepCopyAssign`/`FindParent`, the `<deepCopy><ignore>`/`<stateStorable><exclude>` directives, and the dead `percolationCopyTable`. (`percolation` keeps a slimmed deepCopy trio only because it deep-copies its `class(*) percolationObjects_`, which the generator cannot; its domain-level `solving` flag is kept.) Net ~558 lines deleted.

**Phase 4 — docs, diagnostics, and the two warts.** Documented `recursive="yes"` in the developer guide; extended the #397 abort message to suggest the attribute for bounded cycles; fixed the `objectSet` cache-hijack (via a new `functionClass%isRecursiveShim()` the objectBuilder checks before caching) and the `autoHook` double-registration (the shim's `autoHook` is a no-op); simplified the dead `recursiveConstruct,recursiveSelf` constructor-signature tolerances.

## Design decisions (against §3 of the plan)

- **D1 (kept the opt-in).** `recursive="yes"` remains a one-word opt-in; a non-opted-in cycle still gets the clean #397 construction-time abort rather than a run-time divergence.
- **D2 (family-level shim type).** Option (a) — one generated `<name>Recursive` per family — as recommended.
- **D3 (unified detection).** The build stack now carries the object-under-construction pointer, subsuming the module variables.
- **D4 (deepCopy fix-up).** The audit found that a naïve `copiedSelf` reuse breaks because the generated `deepCopyFinalize` nulls `copiedSelf` top-down — before a shim (whose parent is an ancestor) could read it. The fix was to **stop nulling `copiedSelf` in `deepCopyFinalize`** (it is only ever read inside a deepCopy operation, which always begins with `deepCopyReset` nulling it), so it survives the pass and the shim resolves `recursiveSelf%copiedSelf`. The `parentDeferred` two-pass scheme is retained for the not-yet-copied case.
- **D5 (per-facility semantics).** Implemented as tabulated: forward for methods/state/descriptor/objectType; no-op for `allowedParameters`/`autoHook`; weak, never-counted `recursiveSelf`.

## Two latent bugs found and fixed

1. **Shim deepCopy reference count.** A mold-allocated shim copy starts at `referenceCount=0`; the generated deepCopy resets it to 1 but the shim's did not, so the owner's `objectDestructor` decremented it to −1 and aborted. Surfaced only once the `starFormationHistory` shim was owned by a per-thread-deep-copied-then-destroyed `nodeOperator`.
2. **objectSet cache-hijack** (the second wart from the destruction/memory-leak addendum): the objectBuilder cached the shim in the parameter node, occupying the slot so the real object was never stored and every later requester received the shim — a latent use-after-free. The `isRecursiveShim()` guard restores "real object in the cache".

## Testing

Green on a full build at each phase: generator pin tests, the successful-recursion and #397 abort tests, the three `test-mass-conservation-*` models, the lightcone and fixedAge models, and a new state store/restore round-trip test for the shim (`test-stateRestore-recursion.py`) that reproduces the store run's final tree bit-for-bit after restore.

Relation to #119 summarised in a companion comment there.
